### ISSUE TITLE
Issues with Meta Field value deletion

### ISSUE THREAD

**Issue by prowp** [MinnPost/object-sync-for-salesforce#275]
**Describe the bug**
I'm seeing an issue on a client site where they have a numerical field with data in it. If they delete the data on the salesforce side, just removing the value, the update on the WP side skips, so the old value remains. Example: 20 widgets is the original value, and they remove the 20 from the field and leave it blank in salesforce, the update for that specific field does not happen during the next sync.

**To Reproduce**
1. Add a fieldmap that pulls a field from salesforce to a WP post meta field
2. Add a value for this field in salesforce (I'm not clear if it matters that it's a number field)
3. run a sync to WP
4. See the original value
5. Change the value in Salesforce to blank
6. run a sync to WP
7. Look at the meta value in WP after the udpate, it does not change

**Expected behavior**
The value should be deleted.

Environment
 - WordPress Version: 5.2.2
 - Other relevant plugins Woocommerce
 - PHP Version 7.2

**Additional context**
I tested this by updating to another numerical value, and the update processed. After deleting the value and updating again, the value remained on the WP side after a new sync

**Comment by jonathanstegall** [comment]
@prowp I haven't forgotten about this, but it may be some time before I can devote the time necessary to make sure I can duplicate it, and if so to find a fix for it. It makes sense to me that this would happen. The plugin may be skipping fields with empty values, when in a more correct way, maybe (off the top of my head) it would only skip empty values if it was a newly mapped item.

**Comment by prowp** [comment]
@jonathanstegall Thanks for the follow up. I've instructed the end users to work around this until it can be worked out so it is not urgent. I appreciate the feedback.

**Comment by jonathanstegall** [comment]
I was checking to see if I could quickly find the cause for this. I believe it's [this code](https://github.com/MinnPost/object-sync-for-salesforce/blob/4d3187bb0d9069cf97ae3cceb6592e6186d7a9f0/classes/salesforce_mapping.php#L926-L933) in salesforce_mapping.php.

```
// Make an array because we need to store the methods for each field as well.
if ( isset( $object[ $salesforce_field ] ) && '' !== $object[ $salesforce_field ] ) {
	$params[ $wordpress_field ]          = array();
	$params[ $wordpress_field ]['value'] = $object[ $salesforce_field ];
} else {
	// If we try to save certain fields with empty values, WordPress will silently start skipping stuff. This keeps that from happening.
	continue;
}
```

I do vaguely remember running into the issue where WordPress would silently start skipping that is mentioned in the comment. I also ran a quick test where I emptied out some values in WordPress. They were successfully emptied in Salesforce.

So it appears that, as you say, this issue only happens when values are emptied in Salesforce, and further that it happens because otherwise WordPress was getting confused at some point.

I expect this means we'll have to update all the various update methods inside wordpress.php, maybe by checking to see if a parameter's value is blank and then deleting it if it's a meta field, or emptying it if it's not a meta field. But it could get tricky if it's a required field, even.

**Comment by jonathanstegall** [comment]
When I closed #331 as a duplicate, I forgot to bring over the mention of [this forum post](https://wordpress.org/support/topic/salesforce-null-handling/) which got some recent activity and some code that is possibly worth testing.

**Comment by jonathanstegall** [comment]
Here's another [forum post](https://wordpress.org/support/topic/limiting-contact-pulls-by-custom-sf-value/) about this.

**Comment by jonathanstegall** [comment]
I continue to think that ideally we'd set the value to empty on core tables, and delete the row if it was a (standard) meta field.

Maybe we'd need to create an array of fields to delete.

**Comment by jonathanstegall** [comment]
@prowp I wonder if you'd be able to give me another pair of eyes on the fix for this. See PR #346 or commit 2334b13ab3f4227fee4610a1b83d693e10f68f74 if you'd be willing?

**Comment by jonathanstegall** [comment]
Fixed in #346. We can reopen if it continues, but I am pretty confident it's ok now.
